include/seastar/core/weak_ptr.hh

```diff
@@ -139,6 +139,13 @@ public:
         _ptr_list.push_back(ptr);
         return ptr;
     }
+
+    weak_ptr<const T> weak_from_this() const noexcept {
+        weak_ptr<T> tmp = const_cast<weakly_referencable<T>*>(this)->weak_from_this();
+        weak_ptr<const T> ret(std::exchange(tmp._ptr, nullptr));
+        tmp._hook.swap_nodes(ret._hook);
+        return ret;
+    }
```

Comment:
@xemul hi Pavel, i'd propose a different, but probably more standard compliant, and extensive solution to address this problem.

from the standard, `weak_ptr<Y>` should be constructible from another `weak_ptr<T>`.  as long as `Y` is "compatible" with `T`. see https://eel.is/c++draft/util.sharedptr#util.smartptr.shared.general-6 .

so instead of marking `_ptr_list` mutable. we could just use something like:
```c++
namespace internal {

// quote from
// http://eel.is/c++draft/util.sharedptr#util.smartptr.shared.general-6,
//
// a pointer type Y* is said to be compatible with a pointer type T* when
// either Y* is convertible to T* or Y is U[N] and T is cv U[].
template <typename Y, typename T>
struct bounded_array_of : std::false_type {};

template <typename U, typename T, size_t N>
struct bounded_array_of<U[N], T> : std::is_same<std::remove_cv_t<T>, U[]> {};

template <typename Y, typename T>
concept pointer_type_compatible_with = (std::convertible_to<Y*, T*> ||
                                        bounded_array_of<Y, T>::value);
}
///...
template<typename T>
class weak_ptr {
    template<typename U>
    friend class weakly_referencable;
    template<typename U>
    friend class weak_ptr;
private:
///...
    template<internal::pointer_type_compatible_with<T> Y>
    weak_ptr(const weak_ptr<Y>& o) noexcept {
        if (o._ptr) {
            swap(o._ptr->weak_from_this());
        }
    }
///...
template<typename T>
class weakly_referencable {
///...
    weak_ptr<T> weak_from_this() const noexcept {
        return const_cast<weakly_referencable*>(this)->weak_from_this();
    }
};
```

for couple reasons:

1. more standard compliant. so easier to learn from a regular C++ programmer's perspective
2. we get a better `weak_ptr`, as casting to a compatible pointer type should be possible, and this behavior is more intuitive.
3. less `mutable`. personally, i'd avoid it if possible. as it is an exception, which means.. suprise.

Reply:
`mutable` is not required, I erroneously forgot to remove it. But I like the idea of introducing `weak_ptr<const T>(weak_ptr<T>&&)` constructor, I'll try it.

However at this point no more flexibility is requires from my perspective. In particular, your suggestion means that `weak_ptr<T>(weak_ptr<const T>&&)` will be introduced, since `std::is_convertible<const T, T>::value == true, but it's not correct, casting const pointer to non-const shouldn't happen automatically

Reply:
i think you misread the line of
```c++
std::convertible_to<Y*, T*>
```

also, my original intention was not "flexibility", but more about standard compliance and correctness.

Reply:
> i think you misread the line of
> 
> ```c++
> std::convertible_to<Y*, T*>
> ```

Yes, thank you, I misread the *'s :( However, I'm still concerned about the bounded_array. The

```c++
template <typename Y, typename T>
concept pointer_type_compatible_with = (std::convertible_to<Y*, T*> ||
                                        bounded_array_of<Y, T>::value);
```

means that we can create `weak_ptr<T>` from `weak_ptr<Y>` provided std::converible_to<Y, T>, but in case Y is an array, we cannot get `weak_ptr<Y>` itself, because array cannot inherit from `weakly_referencable`

Reply:
sorry, i was appreciating the the standard library design, and completely ignored the connection between `seastar::weak_ptr` and `seastar::weakly_referencable` . as in our implementation, we don't create a `seastar::weak_ptr` from an existing instance of `seastar::shared_ptr` or `seastar::lw_shared_ptr`, like the standard library does. so we have to to rely on the type of `T` instead of its smart pointer wrapper.